### ISSUE TITLE
1.12 - tests of mpileup and test_vcf_annotate fail - any idea where to look?

### ISSUE THREAD

**Issue by smoe** [samtools/bcftools#1491]
Many thanks!

```
test_vcf_annotate:
	/home/steffen/Med/bcftools/bcftools annotate -a /tmp/CjPTPSdWAX/annotate22.vcf.gz  -c FMT/XX,INFO/XX -x FMT/XX,INFO/XX /tmp/CjPTPSdWAX/annotate22.vcf.gz | /home/steffen/Med/bcftools/bcftools view | grep -v ^##bcftools_

	Non-zero status 1

		Failed to read from standard input: unknown file type


.. failed ...

test_vcf_annotate:
	/home/steffen/Med/bcftools/bcftools annotate -Ob -a /tmp/CjPTPSdWAX/annotate22.vcf.gz  -c FMT/XX,INFO/XX -x FMT/XX,INFO/XX /tmp/CjPTPSdWAX/annotate22.vcf.gz | /home/steffen/Med/bcftools/bcftools view | grep -v ^##bcftools_

	Non-zero status 1

		Failed to read from standard input: unknown file type


.. failed ...
```
and

```
test_mpileup:
	/home/steffen/Med/bcftools/bcftools mpileup -a AD -f /home/steffen/Med/bcftools/test/mpileup/indel-AD.1.fa /home/steffen/Med/bcftools/test/mpileup/indel-AD.1.bam | grep -v ^##bcftools | grep -v ^##reference

	The outputs differ:
		/home/steffen/Med/bcftools/test/mpileup/indel-AD.1.out
		/home/steffen/Med/bcftools/test/mpileup/indel-AD.1.out.new

		[mpileup] 1 samples in 1 input files
		[mpileup] maximum number of reads per input file set to -d 250


			168c168
			< 000000F	537	.	AC	A	0	.	INDEL;IDV=60;IMF=0.48;DP=125;I16=38,30,31,26,215,1051,1927,68083,4080,244800,3420,205200,1555,37389,1281,30547;QS=0.532054,0.467946;VDB=0.0330029;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:68,57
			---
			> 000000F	537	.	AC	A	0	.	INDEL;IDV=60;IMF=0.48;DP=125;I16=38,30,31,26,216,1072,1929,68145,4080,244800,3420,205200,1555,37389,1281,30547;QS=0.531796,0.468204;VDB=0.0330029;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:68,57
			170c170
			< 000000F	538	.	CT	C	0	.	INDEL;IDV=64;IMF=0.512;DP=125;I16=33,34,36,22,205,939,1952,68106,4020,241200,3480,208800,1535,36845,1305,31339;QS=0.524714,0.475286;VDB=0.0281315;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:67,58
			---
			> 000000F	538	.	CT	C	0	.	INDEL;IDV=64;IMF=0.512;DP=125;I16=33,34,36,22,206,950,1952,68106,4020,241200,3480,208800,1535,36845,1305,31339;QS=0.524714,0.475286;VDB=0.0281315;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:67,58
			174c174
			< 000000F	542	.	T	<*>	0	.	DP=124;I16=61,40,0,0,3942,164588,0,0,6060,363600,0,0,2296,55814,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:101,0
			---
			> 000000F	542	.	T	<*>	0	.	DP=124;I16=61,41,0,0,3930,162288,0,0,6120,367200,0,0,2321,56439,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:102,0
			180c180
			< 000000F	548	.	C	<*>	0	.	DP=119;I16=61,41,0,0,3969,163237,0,0,6120,367200,0,0,2412,58868,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:102,0
			---
			> 000000F	548	.	C	<*>	0	.	DP=119;I16=62,41,0,0,3969,162355,0,0,6180,370800,0,0,2437,59493,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:103,0
			184,185c184,185
			< 000000F	552	.	A	<*>	0	.	DP=116;I16=54,47,0,0,3895,160643,0,0,6060,363600,0,0,2349,57103,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:101,0
			< 000000F	553	.	G	<*>	0	.	DP=116;I16=58,46,0,0,4042,167788,0,0,6240,374400,0,0,2424,59158,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:104,0
			---
			> 000000F	552	.	A	<*>	0	.	DP=116;I16=54,48,0,0,3895,157609,0,0,6120,367200,0,0,2374,57728,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:102,0
			> 000000F	553	.	G	<*>	0	.	DP=116;I16=58,47,0,0,4042,165082,0,0,6300,378000,0,0,2449,59783,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:105,0
			199c199
			< 000000F	567	.	G	<*>	0	.	DP=108;I16=56,41,0,0,3823,157549,0,0,5820,349200,0,0,2202,53480,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:97,0
			---
			> 000000F	567	.	G	<*>	0	.	DP=108;I16=56,42,0,0,3823,154597,0,0,5880,352800,0,0,2227,54105,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:98,0
			216c216
			< 000000F	583	.	GAAAAA	GAAAA	0	.	INDEL;IDV=2;IMF=0.0222222;DP=90;I16=47,41,1,1,975,18221,38,722,5280,316800,120,7200,1993,47543,50,1250;QS=0.977475,0.0225252;VDB=5.8801e-07;SGB=-0.453602;MQSB=1;MQ0F=0	PL:AD	0,230,192:88,2
			---
			> 000000F	583	.	GAAAAA	GAAAA	0	.	INDEL;IDV=2;IMF=0.0222222;DP=90;I16=47,41,1,1,977,18253,38,722,5280,316800,120,7200,1993,47543,50,1250;QS=0.977475,0.0225252;VDB=5.8801e-07;SGB=-0.453602;MQSB=1;MQ0F=0	PL:AD	0,230,192:88,2
			224c224
			< 000000F	591	.	G	<*>	0	.	DP=86;I16=40,39,0,0,2985,116439,0,0,4740,284400,0,0,1685,39137,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD0,238,255:79,0
			---
			> 000000F	591	.	G	<*>	0	.	DP=86;I16=40,40,0,0,2985,114981,0,0,4800,288000,0,0,1710,39762,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD0,241,255:80,0
			257c257
			< 000000F	624	.	T	<*>	0	.	DP=45;I16=19,19,0,0,1450,56440,0,0,2280,136800,0,0,742,17306,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD0,114,255:38,0
			---
			> 000000F	624	.	T	<*>	0	.	DP=45;I16=18,20,0,0,1438,55408,0,0,2280,136800,0,0,767,17931,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD0,114,255:38,0
.. failed ...

test_mpileup:
	/home/steffen/Med/bcftools/bcftools mpileup -a AD -f /home/steffen/Med/bcftools/test/mpileup/indel-AD.1.fa -Ob /home/steffen/Med/bcftools/test/mpileup/indel-AD.1.bam | /home/steffen/Med/bcftools/bcftools view  | grep -v ^##bcftools | grep -v ^##reference

	The outputs differ:
		/home/steffen/Med/bcftools/test/mpileup/indel-AD.1.out
		/home/steffen/Med/bcftools/test/mpileup/indel-AD.1.out.new

		[mpileup] 1 samples in 1 input files
		[mpileup] maximum number of reads per input file set to -d 250


			168c168
			< 000000F	537	.	AC	A	0	.	INDEL;IDV=60;IMF=0.48;DP=125;I16=38,30,31,26,215,1051,1927,68083,4080,244800,3420,205200,1555,37389,1281,30547;QS=0.532054,0.467946;VDB=0.0330029;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:68,57
			---
			> 000000F	537	.	AC	A	0	.	INDEL;IDV=60;IMF=0.48;DP=125;I16=38,30,31,26,216,1072,1929,68145,4080,244800,3420,205200,1555,37389,1281,30547;QS=0.531796,0.468204;VDB=0.0330029;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:68,57
			170c170
			< 000000F	538	.	CT	C	0	.	INDEL;IDV=64;IMF=0.512;DP=125;I16=33,34,36,22,205,939,1952,68106,4020,241200,3480,208800,1535,36845,1305,31339;QS=0.524714,0.475286;VDB=0.0281315;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:67,58
			---
			> 000000F	538	.	CT	C	0	.	INDEL;IDV=64;IMF=0.512;DP=125;I16=33,34,36,22,206,950,1952,68106,4020,241200,3480,208800,1535,36845,1305,31339;QS=0.524714,0.475286;VDB=0.0281315;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:67,58
			174c174
			< 000000F	542	.	T	<*>	0	.	DP=124;I16=61,40,0,0,3942,164588,0,0,6060,363600,0,0,2296,55814,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:101,0
			---
			> 000000F	542	.	T	<*>	0	.	DP=124;I16=61,41,0,0,3930,162288,0,0,6120,367200,0,0,2321,56439,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:102,0
			180c180
			< 000000F	548	.	C	<*>	0	.	DP=119;I16=61,41,0,0,3969,163237,0,0,6120,367200,0,0,2412,58868,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:102,0
			---
			> 000000F	548	.	C	<*>	0	.	DP=119;I16=62,41,0,0,3969,162355,0,0,6180,370800,0,0,2437,59493,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:103,0
			184,185c184,185
			< 000000F	552	.	A	<*>	0	.	DP=116;I16=54,47,0,0,3895,160643,0,0,6060,363600,0,0,2349,57103,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:101,0
			< 000000F	553	.	G	<*>	0	.	DP=116;I16=58,46,0,0,4042,167788,0,0,6240,374400,0,0,2424,59158,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:104,0
			---
			> 000000F	552	.	A	<*>	0	.	DP=116;I16=54,48,0,0,3895,157609,0,0,6120,367200,0,0,2374,57728,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:102,0
			> 000000F	553	.	G	<*>	0	.	DP=116;I16=58,47,0,0,4042,165082,0,0,6300,378000,0,0,2449,59783,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:105,0
			199c199
			< 000000F	567	.	G	<*>	0	.	DP=108;I16=56,41,0,0,3823,157549,0,0,5820,349200,0,0,2202,53480,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:97,0
			---
			> 000000F	567	.	G	<*>	0	.	DP=108;I16=56,42,0,0,3823,154597,0,0,5880,352800,0,0,2227,54105,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:98,0
			216c216
			< 000000F	583	.	GAAAAA	GAAAA	0	.	INDEL;IDV=2;IMF=0.0222222;DP=90;I16=47,41,1,1,975,18221,38,722,5280,316800,120,7200,1993,47543,50,1250;QS=0.977475,0.0225252;VDB=5.8801e-07;SGB=-0.453602;MQSB=1;MQ0F=0	PL:AD	0,230,192:88,2
			---
			> 000000F	583	.	GAAAAA	GAAAA	0	.	INDEL;IDV=2;IMF=0.0222222;DP=90;I16=47,41,1,1,977,18253,38,722,5280,316800,120,7200,1993,47543,50,1250;QS=0.977475,0.0225252;VDB=5.8801e-07;SGB=-0.453602;MQSB=1;MQ0F=0	PL:AD	0,230,192:88,2
			224c224
			< 000000F	591	.	G	<*>	0	.	DP=86;I16=40,39,0,0,2985,116439,0,0,4740,284400,0,0,1685,39137,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD0,238,255:79,0
			---
			> 000000F	591	.	G	<*>	0	.	DP=86;I16=40,40,0,0,2985,114981,0,0,4800,288000,0,0,1710,39762,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD0,241,255:80,0
			257c257
			< 000000F	624	.	T	<*>	0	.	DP=45;I16=19,19,0,0,1450,56440,0,0,2280,136800,0,0,742,17306,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD0,114,255:38,0
			---
			> 000000F	624	.	T	<*>	0	.	DP=45;I16=18,20,0,0,1438,55408,0,0,2280,136800,0,0,767,17931,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD0,114,255:38,0
.. failed ...

test_mpileup:
	/home/steffen/Med/bcftools/bcftools mpileup -a AD -f /home/steffen/Med/bcftools/test/mpileup/indel-AD.1.fa /home/steffen/Med/bcftools/test/mpileup/indel-AD.1.cram | grep -v ^##bcftools | grep -v ^##reference

	The outputs differ:
		/home/steffen/Med/bcftools/test/mpileup/indel-AD.1.out
		/home/steffen/Med/bcftools/test/mpileup/indel-AD.1.out.new

		[mpileup] 1 samples in 1 input files
		[mpileup] maximum number of reads per input file set to -d 250


			168c168
			< 000000F	537	.	AC	A	0	.	INDEL;IDV=60;IMF=0.48;DP=125;I16=38,30,31,26,215,1051,1927,68083,4080,244800,3420,205200,1555,37389,1281,30547;QS=0.532054,0.467946;VDB=0.0330029;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:68,57
			---
			> 000000F	537	.	AC	A	0	.	INDEL;IDV=60;IMF=0.48;DP=125;I16=38,30,31,26,216,1072,1929,68145,4080,244800,3420,205200,1555,37389,1281,30547;QS=0.531796,0.468204;VDB=0.0330029;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:68,57
			170c170
			< 000000F	538	.	CT	C	0	.	INDEL;IDV=64;IMF=0.512;DP=125;I16=33,34,36,22,205,939,1952,68106,4020,241200,3480,208800,1535,36845,1305,31339;QS=0.524714,0.475286;VDB=0.0281315;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:67,58
			---
			> 000000F	538	.	CT	C	0	.	INDEL;IDV=64;IMF=0.512;DP=125;I16=33,34,36,22,206,950,1952,68106,4020,241200,3480,208800,1535,36845,1305,31339;QS=0.524714,0.475286;VDB=0.0281315;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:67,58
			174c174
			< 000000F	542	.	T	<*>	0	.	DP=124;I16=61,40,0,0,3942,164588,0,0,6060,363600,0,0,2296,55814,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:101,0
			---
			> 000000F	542	.	T	<*>	0	.	DP=124;I16=61,41,0,0,3930,162288,0,0,6120,367200,0,0,2321,56439,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:102,0
			180c180
			< 000000F	548	.	C	<*>	0	.	DP=119;I16=61,41,0,0,3969,163237,0,0,6120,367200,0,0,2412,58868,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:102,0
			---
			> 000000F	548	.	C	<*>	0	.	DP=119;I16=62,41,0,0,3969,162355,0,0,6180,370800,0,0,2437,59493,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:103,0
			184,185c184,185
			< 000000F	552	.	A	<*>	0	.	DP=116;I16=54,47,0,0,3895,160643,0,0,6060,363600,0,0,2349,57103,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:101,0
			< 000000F	553	.	G	<*>	0	.	DP=116;I16=58,46,0,0,4042,167788,0,0,6240,374400,0,0,2424,59158,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:104,0
			---
			> 000000F	552	.	A	<*>	0	.	DP=116;I16=54,48,0,0,3895,157609,0,0,6120,367200,0,0,2374,57728,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:102,0
			> 000000F	553	.	G	<*>	0	.	DP=116;I16=58,47,0,0,4042,165082,0,0,6300,378000,0,0,2449,59783,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:105,0
			199c199
			< 000000F	567	.	G	<*>	0	.	DP=108;I16=56,41,0,0,3823,157549,0,0,5820,349200,0,0,2202,53480,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:97,0
			---
			> 000000F	567	.	G	<*>	0	.	DP=108;I16=56,42,0,0,3823,154597,0,0,5880,352800,0,0,2227,54105,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD	0,255,255:98,0
			216c216
			< 000000F	583	.	GAAAAA	GAAAA	0	.	INDEL;IDV=2;IMF=0.0222222;DP=90;I16=47,41,1,1,975,18221,38,722,5280,316800,120,7200,1993,47543,50,1250;QS=0.977475,0.0225252;VDB=5.8801e-07;SGB=-0.453602;MQSB=1;MQ0F=0	PL:AD	0,230,192:88,2
			---
			> 000000F	583	.	GAAAAA	GAAAA	0	.	INDEL;IDV=2;IMF=0.0222222;DP=90;I16=47,41,1,1,977,18253,38,722,5280,316800,120,7200,1993,47543,50,1250;QS=0.977475,0.0225252;VDB=5.8801e-07;SGB=-0.453602;MQSB=1;MQ0F=0	PL:AD	0,230,192:88,2
			224c224
			< 000000F	591	.	G	<*>	0	.	DP=86;I16=40,39,0,0,2985,116439,0,0,4740,284400,0,0,1685,39137,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD0,238,255:79,0
			---
			> 000000F	591	.	G	<*>	0	.	DP=86;I16=40,40,0,0,2985,114981,0,0,4800,288000,0,0,1710,39762,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD0,241,255:80,0
			257c257
			< 000000F	624	.	T	<*>	0	.	DP=45;I16=19,19,0,0,1450,56440,0,0,2280,136800,0,0,742,17306,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD0,114,255:38,0
			---
			> 000000F	624	.	T	<*>	0	.	DP=45;I16=18,20,0,0,1438,55408,0,0,2280,136800,0,0,767,17931,0,0;QS=1,0;MQSB=1;MQ0F=0	PL:AD0,114,255:38,0
.. failed ...

test_mpileup:
	/home/steffen/Med/bcftools/bcftools mpileup -a AD -f /home/steffen/Med/bcftools/test/mpileup/indel-AD.1.fa -Ob /home/steffen/Med/bcftools/test/mpileup/indel-AD.1.cram | /home/steffen/Med/bcftools/bcftools view  | grep -v ^##bcftools | grep -v ^##reference

	The outputs differ:
		/home/steffen/Med/bcftools/test/mpileup/indel-AD.1.out
		/home/steffen/Med/bcftools/test/mpileup/indel-AD.1.out.new

		[mpileup] 1 samples in 1 input files
		[mpileup] maximum number of reads per input file set to -d 250


			168c168
			< 000000F	537	.	AC	A	0	.	INDEL;IDV=60;IMF=0.48;DP=125;I16=38,30,31,26,215,1051,1927,68083,4080,244800,3420,205200,1555,37389,1281,30547;QS=0.532054,0.467946;VDB=0.0330029;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:68,57
			---
			> 000000F	537	.	AC	A	0	.	INDEL;IDV=60;IMF=0.48;DP=125;I16=38,30,31,26,216,1072,1929,68145,4080,244800,3420,205200,1555,37389,1281,30547;QS=0.531796,0.468204;VDB=0.0330029;SGB=-0.693147;MQSB=1;MQ0F=0	PL:AD	255,0,255:68,57
```

**Comment by jmarshall** [comment]
I see Debian has begun to package bcftools 1.12, but has not yet updated to htslib 1.12.

_bcftools/debian/control_:
```
Source: bcftools
Maintainer: Debian Med Packaging Team <debian-med-packaging@lists.alioth.debian.org>
[…]
Build-Depends: […]
               libhts-dev (>= 1.11),
```

**Comment by pd3** [comment]
@smoe Does this solve the problem?

**Comment by smoe** [comment]
@pd3 , many thanks, I'll investigate, having missed your first reply.

**Comment by smoe** [comment]
I can confirm. With htslib updated to 1.12 it just builds and tests nicely. Many thanks. And - also sorry for me not having checked this myself - I was somehow not aware of that dependency.

Since this was not a runtime error, would there be a point in checking the version that bcftools are linked against and write a warning to stderr when versions don't match?

**Comment by jmarshall** [comment]
As you saw, bcftools 1.12 compiles and functions with htslib 1.11, so it would not be accurate to enforce a dependency on htslib >= 1.12.

OTOH there are bugs in htslib 1.11 that have been fixed in htslib 1.12, and that of course has an effect on bcftools's behaviour. Hence unsurprisingly bcftools's test suite had failures where the difference in libhts behaviour was reflected in bcftools output. So it might be worth emitting a warning from `make test` (i.e., in _test/test.pl_ or elsewhere in the test suite) when htslib is older than bcftools to provide a clue in this scenario.

Note that to date htslib, samtools, and bcftools releases have all occurred in lockstep. However one of the reasons for separating them was so that samtools or bcftools could make additional releases if desired. Naively comparing version numbers would interfere with this.
